OneSignal/OneSignal-iOS-SDK
# [5.0.0] Allow the SDK the chance to recover from failed Create User

# Description
## One Line Summary
If we receive a non-retryable error response to Create User, we will keep it cached, pause the operation repo, and retry on a new session.

## Details

### Motivation
A failed Create User request would leave the SDK in a bad state. Perhaps subscriptions or tags are over limit (not possible for creating Guest User but may be possible for logging into an identified user). To allow the SDK a chance to recover, don't remove the request from cache and pause the operation repo. Then we will retry this request on a new session

### Details of Commits
* If a createUser fails, let's pause the operation repo. We will still poll in case the operation repo becomes unpaused, but we won't flush if it is paused.
* Fix a bug: after a create user request is uncached, update the `originalPushToken` property to be the push token associated with this request. Because, we update the push subscription before sending this request so the token sent may be different than the token the request had at the time of caching (common for it to start `null` and get updated by the time of the request).
* On a new session, we will trigger the user executor to execute any pending requests, in case there is a failed create user request that should be retried. We can't rely on the operation repo flushing to trigger the user executor as it bypasses the operation repo. We also unpause the operation repo to effectively restart it.
* Add **timezone_id** to property model and send it in the User Create. This was previously missed, and more work will need to be done if we want to detect changes to `timezone` and update the model.

# Testing
## Unit testing
* No new unit tests added but existing tests are updated to pass

## Manual testing
iPhone 13 device on version 16.6
1. Force the client to return a `409` to the Create User request.
2. Use a fresh app install
3. See the request fail, the operation repo be paused
4. Background the app and reopen to trigger a new session
5. See the operation repo unpause and the failed Create User request try again
6. If it succeeds, hydration of onesignal_id and subscription_id happens correctly.
7. If it fails again, goes back to step 3

# Affected code checklist

# Checklist
## Overview
     - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR

## Testing

## Final pass
      - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
      - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/OneSignal/OneSignal-iOS-SDK/1290)
